### PR TITLE
Typed Transactions Part 2: Protocol Schedule Based

### PR DESCRIPTION
**Pull request by RatanRSur**
In order to get the transaction decoding logic to switch over on network upgrades and reject future transactions, the new `ProtocolRLPSpec` needs to be accessible by ProtocolSchedule. Also, because RLP (de)serialization is no longer static on the data type, I've been moving it out of the data type classes and into `ProtocolRLPSpec`. Encoding is static because it's assumed to be backwards-compatible and we trust the objects we've already parsed to be valid for all uses.

Decoders are also assumed to be backwards compatible so we can use them without issue in decoding RLP on our disk.

Basically, the only things we want to be wary of are things that are gossiped to us or that we get through the rpc."

Full syncing mainnet just to make sure nothing weird happens at fork block boundaries:
https://grafana-metrics-ohio.ops.pegasys.tech/d/5S-6O8VZk/besu-overview?orgId=1&var-system=dev-besu-ohio-mainnet-typed-tx-sync

### CODE REVIEW

File: consensus/ibft/src/main/java/org/hyperledger/besu/consensus/ibft/messagewrappers/Proposal.java [hunk 1]
```diff
@@ -74,7 +77,11 @@ public static Proposal decode(final Bytes data) {
     final RLPInput rlpIn = RLP.input(data);
     rlpIn.enterList();
     final SignedData<ProposalPayload> payload = SignedData.readSignedProposalPayloadFrom(rlpIn);
-    final Block proposedBlock = Block.readFrom(rlpIn, IbftBlockHeaderFunctions.forCommittedSeal());
+    final Block proposedBlock =
+        ProtocolRLPSpec.decodeBlockStandalone(
+            IbftProtocolSchedule.create(new StubGenesisConfigOptions()),
```
Comment (shemnon) [inline consensus/ibft/src/main/java/org/hyperledger/besu/consensus/ibft/messagewrappers/Proposal.java:82]: Creating a stub genesis during a generic decode seems like (a) memory thrash for performance and (b) like we will lose config data if we depend on that for the object.  It feels like concerns are mixed improperly if we need to do this.

File: consensus/ibft/src/main/java/org/hyperledger/besu/consensus/ibft/messagewrappers/RoundChange.java [hunk 2]
```diff
@@ -73,7 +76,12 @@ public static RoundChange decode(final Bytes data) {
         SignedData.readSignedRoundChangePayloadFrom(rlpIn);
     Optional<Block> block = Optional.empty();
     if (!rlpIn.nextIsNull()) {
-      block = Optional.of(Block.readFrom(rlpIn, IbftBlockHeaderFunctions.forCommittedSeal()));
+      block =
+          Optional.of(
+              ProtocolRLPSpec.decodeBlockStandalone(
+                  IbftProtocolSchedule.create(new StubGenesisConfigOptions()),
```
Comment (shemnon) [inline consensus/ibft/src/main/java/org/hyperledger/besu/consensus/ibft/messagewrappers/RoundChange.java:82]: What if the IBFT network forks between versions that have typed transactions and those that don't?  If we are passing in the same freshly created schedule then do we need to pass it in?  Or create it every time?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/encoding/ProtocolRLPSpec.java [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+  interface Decoder<T> {
+    T decode(RLPInput input);
+  }
+
+  ImmutableMap<TransactionType, Encoder<Transaction>> TYPED_TRANSACTION_ENCODERS =
+      ImmutableMap.of(TransactionType.EIP1559, ProtocolRLPSpec::encodeEIP1559);
+
+  // This is all going to change in the next PR because we had to settle some spec ambiguities on a
+  // breakout room call
+  static void encode(final Transaction transaction, final RLPOutput rlpOutput) {
```
Comment (shemnon) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/encoding/ProtocolRLPSpec.java:61]: I don't like how the `writeTo` and `readFrom` methods were removed from the associated data objects.  It forces these encoding objects to access what should be encapsulated data.  

It also forces us to do weird lambda dances, where `foo::writeTo` becomes `rlpOut -> Static.method(foo, rlpOut)`.  This transform makes me thing we did something wrong.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/BodyValidation.java [hunk 4]
```diff
@@ -54,9 +56,13 @@ private static Bytes indexKey(final int i) {
   public static Hash transactionsRoot(final List<Transaction> transactions) {
     final MerklePatriciaTrie<Bytes, Bytes> trie = trie();
 
-    for (int i = 0; i < transactions.size(); ++i) {
-      trie.put(indexKey(i), RLP.encode(transactions.get(i)::writeTo));
-    }
+    IntStream.range(0, transactions.size())
```
Comment (shemnon) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/BodyValidation.java:59]: a for loop will have better performance in this instance, and is more readable.  JVM JITs have been recognizing this pattern for decades and produce very nice code for it.

File: ethereum/core/src/test-support/java/org/hyperledger/besu/ethereum/core/InMemoryStorageProvider.java [hunk 5]
```diff
@@ -38,15 +39,19 @@
 public class InMemoryStorageProvider implements StorageProvider {
 
   public static MutableBlockchain createInMemoryBlockchain(final Block genesisBlock) {
-    return createInMemoryBlockchain(genesisBlock, new MainnetBlockHeaderFunctions());
+    return createInMemoryBlockchain(
+        genesisBlock, MainnetProtocolSchedule.DEFAULT, new MainnetBlockHeaderFunctions());
```
Comment (shemnon) [inline ethereum/core/src/test-support/java/org/hyperledger/besu/ethereum/core/InMemoryStorageProvider.java:43]: The protocol schedule should be an argument, not a presumed default.  We will want to have test cases on both sides of the typed transactions fork.

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/task/GetHeadersFromPeerByNumberTask.java [hunk 6]
```diff
@@ -28,19 +28,17 @@
 public class GetHeadersFromPeerByNumberTask extends AbstractGetHeadersFromPeerTask {
   private static final Logger LOG = LogManager.getLogger();
 
-  private final long blockNumber;
```
Comment (shemnon) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/task/GetHeadersFromPeerByNumberTask.java:31]: Are these changes essential to Typed Transactions?

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/FastDownloaderFactory.java [hunk 7]
```diff
@@ -57,12 +57,13 @@
...
     final Path fastSyncDataDirectory = dataDirectory.resolve(FAST_SYNC_FOLDER);
     final FastSyncStateStorage fastSyncStateStorage =
-        new FastSyncStateStorage(fastSyncDataDirectory);
+        new FastSyncStateStorage(fastSyncDataDirectory, protocolSchedule);
 
     if (syncConfig.getSyncMode() != SyncMode.FAST) {
       if (fastSyncStateStorage.isFastSyncInProgress()) {
         throw new IllegalStateException(
-            "Unable to change the sync mode when fast sync is incomplete, please restart with fast sync mode");
+            "Unable to change the sync mode when fast sync is incomplete, please restart with fast"
```
Comment (shemnon) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/FastDownloaderFactory.java:65]: Is it necessary to split these strings into 120 column bits?  It increases the line count in an already large PR.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetProtocolSchedule.java [hunk 8]
```diff
@@ -28,10 +28,8 @@
 
   public static final BigInteger DEFAULT_CHAIN_ID = BigInteger.ONE;
 
-  public static ProtocolSchedule create() {
-    return fromConfig(
-        GenesisConfigFile.mainnet().getConfigOptions(), PrivacyParameters.DEFAULT, false);
-  }
+  public static final ProtocolSchedule DEFAULT =
```
Comment (shemnon) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetProtocolSchedule.java:31]: Changing from a static method to a constant is inessential to Type Transactions and has a long reach, adding many unneeded files and diff lines.

File: config/src/main/java/org/hyperledger/besu/config/GenesisConfigFile.java [hunk 9]
```diff
@@ -75,6 +79,25 @@ public static GenesisConfigFile fromConfig(final ObjectNode config) {
     return new GenesisConfigFile(normalizeKeys(config));
   }
 
+  public static GenesisConfigOptions getMainnetConfigOptions() {
```
Comment (shemnon) [inline config/src/main/java/org/hyperledger/besu/config/GenesisConfigFile.java:82]: This and constant is supports is inessential to Typed Transactions and should be in it's own PR.

**Review by shemnon — COMMENTED** [review]
This is a large (net 2300 line) PR and there are many changes that are not essential to Type Transactions.  Please move the inessential changes to their own PRs.

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/FastDownloaderFactory.java (hunk 7 above)
Comment (RatanRSur) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/FastDownloaderFactory.java:65]: This is an artifact of me running google-java-format externally to intellij sometimes as a git hook, apparently I don't have it set  up with the same settings as intellij. I'll remove them.

File: config/src/main/java/org/hyperledger/besu/config/GenesisConfigFile.java (hunk 9 above)
Comment (shemnon) [reply config/src/main/java/org/hyperledger/besu/config/GenesisConfigFile.java:82]: Should this be deleted?  Didn't it get moved here - https://github.com/hyperledger/besu/pull/1688/files#diff-df7c51531e08d1f892222e589588a0f67d907b10d6f4d84f6427eacb46ee4a63R38?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/chain/DefaultBlockchain.java [hunk 10]
```diff
@@ -532,7 +532,8 @@ private void setGenesis(final Block genesisBlock) {
       if (!genesisHash.get().equals(genesisBlock.getHash())) {
         throw new InvalidConfigurationException(
             "Supplied genesis block does not match stored chain data.\n"
-                + "Please specify a different data directory with --data-path or specify the original genesis file with --genesis-file.");
+                + "Please specify a different data directory with --data-path or specify the"
```
Comment (shemnon) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/chain/DefaultBlockchain.java:535]: This edit is non-essential.
